mlir/unittests/Pass/PassManagerTest.cpp

```diff
@@ -181,4 +182,48 @@ TEST(PassManagerTest, PassInitialization) {
   EXPECT_TRUE(succeeded(pm.run(module.get())));
 }
 
+struct ReRegisterPass
+    : public PassWrapper<ReRegisterPass, OperationPass<ModuleOp>> {
+  MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(ReRegisterPass)
+
+  ReRegisterPass(std::string annotation) : annotation(annotation) {}
+
+  std::string annotation;
+
+  void runOnOperation() override {
+    ModuleOp op = this->getOperation();
+    Builder builder(op);
+    op->walk([this, &builder](Operation *op) {
+      op->setAttr(annotation, builder.getUnitAttr());
+    });
+  }
+};
+
+TEST(PassManagerTest, PassReRegistration) {
+  MLIRContext context;
+  context.allowUnregisteredDialects();
+
+  std::string moduleStr = R"mlir(
+    module {
+      "custom.op1"() : () -> ()
+      "custom.op2"() : () -> ()
+    }
+  )mlir";
+
+  OwningOpRef<ModuleOp> module =
+      parseSourceString<ModuleOp>(moduleStr, &context);
+
+  // Instantiate and run pass in first configuration.
+  auto pm = PassManager::on<ModuleOp>(&context);
+  pm.addPass(std::make_unique<ReRegisterPass>("custom.first"));
+  EXPECT_TRUE(succeeded(pm.run(module.get())));
+  module->walk([](Operation *op) { EXPECT_TRUE(op->hasAttr("custom.first")); });
+
+  // Adding a "reconfiguration" of the pass, i.e., with a different annotation.
+  pm.addPass(std::make_unique<ReRegisterPass>("custom.second"));
```

Comment:
1) I don't quite get where there is a "registration" happening here?
2) Why isn't this a pass option?

Reply:
This is only an example of course for the sake of the rest - a more complicated use case would pass more complex data structures to the constructor that couldn't be passed as at the CLI.

Reply:
Also the reregistration happens because ultimately this calls `registerPass` again.

Reply:
> This is only an example of course for the sake of the rest - a more complicated use case would pass more complex data structures to the constructor that couldn't be passed as at the CLI.

I'm not sure I follow actually: how do you run such pass with `mlir-opt`?

> Also the reregistration happens because ultimately this calls registerPass again.

Why does the registration part matters in this test? Is there an assertion hit if we run this test without the change in the library?

Reply:
> I'm not sure I follow actually: how do you run such pass with `mlir-opt`?

You can't but `mlir-opt` isn't the only way that passes are run right? Anyone that embeds the pass manager in a library can register passes dynamically.

> Why does the registration part matters in this test? Is there an assertion hit if we run this test without the change in the library?

No but without the patch the second `addPass -> registerPass` is a no-op because of `try_emplace` which doesn't update the value (the pass info) associated with the key (the pass arg) if the key has already been stored in `passRegistry`.

Reply:
> You can't but mlir-opt isn't the only way that passes are run right?

I actually use it as a litmus test for "is this a correct setup".

> No but without the patch the second addPass -> registerPass is a no-op because of try_emplace which doesn't update the value (the pass info) associated with the key (the pass arg) if the key has already been stored in passRegistry.

I understand that, but I have yet to understand why it is a problem?
If you write a test, I would expect it in general to fail without the other changes, otherwise it's not really testing the change.

Reply:
Now maybe I'm missing something - it will fail because without the change. If you run the pass manager again, even after adding the "second" pass, you will not see `custom.second` attributes in the second walk right? I mean this isn't a question, I know I tested with and without locally.

Reply:
Ah sorry sorry you're right. One sec.

Reply:
Okay all set (at least as far as actually exercising the code). Sorry I conflated "adding a pass to a pass manager" and "registering a pass" and so you were right the first test didn't actually "register" anything.

Reply:
Regarding

> I actually use it as a litmus test for "is this a correct setup".

Maybe it's not. I guess one could always do what I'm imagining by just registering the same pass under two different args but I'm not sure that's correct either? 

The particular use case I have in mind is a pass that has an analysis that it runs, where there's a default "analyzer" provided but alternatively a user (that calls the pass manager through a library call) can provide their own (i.e., by passing the analyzer to the pass constructor).

Reply:
> . I guess one could always do what I'm imagining by just registering the same pass under two different args but I'm not sure that's correct either?

That seems fine: that would work with `mlir-opt` right?

Reply:
If you really think this is bad practice then sure I'll workaround but I think my case for this change is like this:

1. There are library users of `passRegistry` (the Python bindings at minimum);
2. The current behavior is counter-intuitive (or a bug) in that re-registering is a no-op but there's no warming/error/message that it's a no-op;
3. This change is completely fail-safe because of the immediately subsequent typeid check.

Reply:
I'm still not sure what's the problem or how you're change here is solving it in a principled way?
One registration over-writing silently the previous one does not seem better to me than ignoring the second one. Actually I'd say it is worse: it would make your example work, so someone may think it is fine, when it turns out they over-wrote something. Better have it just "not work".

The problem is that writing such a pass that isn't "hermetic" is not desirable. But I don't see how we can protect people against themselves? That is: it is C++ and folks have access to every low-level aspects of the language.

Reply:
It looks like the problem this is trying to solve is this:

> The current behavior is counter-intuitive (or a bug) in that re-registering is a no-op but there's no warming/error/message that it's a no-op;

Do I understand correctly that registering a different constructor function for the same pass name + TypeID is currently allowed? This sounds undesirable to me, because it may lead to surprises, for example having two `PassRegistration` objects that register the same pass with different configuration may end up having the pass behavior depend on the order of construction of these objects, which may in turn depend on the order of linking.

That being said, I don't think this PR really solves that problem. Instead of prioritizing the first registered constructor, it will prioritize the last registered one. But ordering is still an issue.

I would rather consider removing the custom constructor callback for the pass. All pass classes could provide a (C++) default constructor, so re-registering a pass always points to the same constructor. This may be quite brutal to downstreams though...

Reply:
Right now the pass registration contract with the user is that they provide a `PassAllocatorFunction` callback : `using PassAllocatorFunction = std::function<std::unique_ptr<Pass>()>;`

From there the system will create the pass and check that the pass name  isn't registered yet, and if it is that the TypeID matches.

To forbid this we would have to change the API here, maybe expose exclusively a `template<typename PassTy> registerPass();` and nothing else?